CoseSignTool/CoseSign1/DetachedSignatureFactory.cs

```diff
@@ -0,0 +1,301 @@
+    /// </summary>
+    public HashAlgorithm HashAlgorithm => InternalHashAlgorithm;
+
+    /// <summary>
+    /// The HashAlgorightmName this factory is using.
+    /// </summary>
+    public HashAlgorithmName HashAlgorithmName => InternalHashAlgorithmName;
+
+    /// <summary>
+    /// The CoseSign1 Message Factory this factory is using.
+    /// </summary>
+    public ICoseSign1MessageFactory MessageFactory => InternalMessageFactory;
+
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="DetachedSignatureFactory"/> class using the <see cref="HashAlgorithmName.SHA256"/> hash algorithm and a <see cref="CoseSign1MessageFactory"/>.
+    /// </summary>
+    public DetachedSignatureFactory() : this(HashAlgorithmName.SHA256)
+    {
+    }
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="DetachedSignatureFactory"/> class using the specified hash algorithm and a <see cref="CoseSign1MessageFactory"/>.
+    /// </summary>
+    /// <param name="hashAlgorithmName">The hashing algorithm name to be used when performing hashing operations.</param>
+    public DetachedSignatureFactory(HashAlgorithmName hashAlgorithmName) : this(hashAlgorithmName, new CoseSign1MessageFactory())
+    {
+    }
+
+    /// <summary>
+    /// Creates a new instance of the <see cref="DetachedSignatureFactory"/> class using the specified hash algorithm and the specified <see cref="ICoseSign1MessageFactory"/>.
+    /// </summary>
+    /// <param name="hashAlgorithmName">The hashing algorithm name to be used when performing hashing operations.</param>
+    /// <param name="coseSign1MessageFactory">The CoseSign1MessageFactory to be used when creating COSE Sign1 messages.</param>
+    public DetachedSignatureFactory(HashAlgorithmName hashAlgorithmName, ICoseSign1MessageFactory coseSign1MessageFactory)
+    {
+        InternalHashAlgorithmName = hashAlgorithmName;
+        InternalHashAlgorithm = HashAlgorithm.Create(hashAlgorithmName.Name);
+        InternalMessageFactory = coseSign1MessageFactory;
+    }
+
+    /// <summary>
+    /// Creates a detached signature of the specified payload returned as a <see cref="CoseSign1Message"/> following the rules in this class description.
+    /// </summary>
+    /// <param name="payload">The payload to create a detached signature for.</param>
+    /// <param name="signingKeyProvider">The COSE signing key provider to be used for the signing operation within the <see cref="ICoseSign1MessageFactory"/>.</param>
+    /// <param name="contentType">A media type string following https://datatracker.ietf.org/doc/html/rfc6838.</param>
+    /// <returns>A COSE Sign1 Message which can be used as a detached signature validation of the payload.</returns>
+    /// <exception cref="ArgumentNullException">The contentType parameter was empty or null</exception>
+    public CoseSign1Message CreateDetachedSignature(
+        ReadOnlyMemory<byte> payload,
+        ICoseSigningKeyProvider signingKeyProvider,

```

Comment:
Would it make sense for signingKeyProvider or contentType to have default values?